pg_search/src/postgres/customscan/pdbscan/mod.rs

```diff
@@ -1321,172 +1315,27 @@ unsafe fn is_partitioned_table_setup(
-                    rte_index,
-                    columns,
-                );
-            }
-        }
-
-        pg_sys::NodeTag::T_BoolExpr => {
-            let boolexpr = node.cast::<pg_sys::BoolExpr>();
-            let args = PgList::<pg_sys::Node>::from_pg((*boolexpr).args);
-            for arg in args.iter_ptr() {
-                collect_maybe_fast_field_vars_from_quals(
-                    arg as *mut pg_sys::Node,
-                    rte_index,
-                    columns,
-                );
-            }
-        }
-
-        // Handle function calls
-        pg_sys::NodeTag::T_FuncExpr => {
-            let funcexpr = node.cast::<pg_sys::FuncExpr>();
-            let args = PgList::<pg_sys::Node>::from_pg((*funcexpr).args);
-            for arg in args.iter_ptr() {
-                collect_maybe_fast_field_vars_from_quals(
-                    arg as *mut pg_sys::Node,
-                    rte_index,
-                    columns,
-                );
-            }
-        }
-
-        // Direct Var references
-        pg_sys::NodeTag::T_Var => {
-            let var = node.cast::<pg_sys::Var>();
-            if (*var).varno as u32 == rte_index {
-                columns.insert((*var).varattno);
-            }
-        }
-
-        // Default - we don't handle other node types for simplicity
-        _ => {}
-    }
-}
-
 /// Gather all columns referenced by the specified RTE (Range Table Entry) throughout the query.
 /// This gives us a more complete picture than just looking at the target list.
 ///
-/// This function is critical for issue #2505 where we need to detect all columns used in JOIN
+/// This function is critical for issue #2505/#2556 where we need to detect all columns used in JOIN
 /// conditions to ensure we select the right execution method. Previously, only looking at the
 /// target list would miss columns referenced in JOIN conditions, leading to execution-time errors.
-///
-/// However, it skips full-text search operators because they're not relevant to our
-/// execution method selection.
 unsafe fn collect_maybe_fast_field_referenced_columns(
-    root: *mut pg_sys::PlannerInfo,
     rte_index: pg_sys::Index,
+    rel: *mut pg_sys::RelOptInfo,
 ) -> HashSet<pg_sys::AttrNumber> {
     let mut referenced_columns = HashSet::new();
 
-    // First check the target list (columns in SELECT)
-    let target_list = (*(*root).parse).targetList;
-    let tlist = PgList::<pg_sys::TargetEntry>::from_pg(target_list);
-
-    // Process all Var nodes in the target list
-    for te in tlist.iter_ptr() {
-        if let Some(var) = nodecast!(Var, T_Var, (*te).expr) {
+    // Check reltarget
+    let reltarget_exprs = PgList::<pg_sys::Expr>::from_pg((*(*rel).reltarget).exprs);
+    for rte in reltarget_exprs.iter_ptr() {
+        if let Some(var) = nodecast!(Var, T_Var, rte) {
             if (*var).varno as u32 == rte_index {
                 referenced_columns.insert((*var).varattno);
             }
         }
```

Comment:
I ran the test suite with a warning/log, and the only other type I encountered here was `PlaceHolderVar`: will spend a bit more time looking at that and add a comment.

@mdashti and I discussed the fact that it is possible that `(*(*rel).reltarget).exprs` is what we should be inspecting _instead_ of the `(*(*root).parse).targetList`, because it seems to have been simplified and expanded.